### PR TITLE
Add support for sending the password for a share by Nextcloud Talk

### PR DESCRIPTION
**Pull request by danxuliu**
Talk part of nextcloud/server#10238

When a file is shared by email the file can be accessed by anyone as long as the token of the share is known, so to limit that access the share can be protected with a password. When a password was set it was also sent by email to the sharee. This pull request (and its server counterpart) adds another option: sending the password with Nextcloud Talk.

Now when the sharer protects the share with a password no email is sent. Instead, when the receiver tries to open the share the public share authentication page is shown, which now contains a _Request password_ button. When that button is clicked an embedded Nextcloud UI is shown and a conversation is started between the sharer and the user trying to open the file, which allows the sharer to verify the identity of the user before telling her the password and thus granting her access to the file.

Note that it is the sharer who decides whether the receiver of an email share can call her or not; you will not be bothered with password requests if you do not choose to ;-)

This pull request provides the handling of rooms and the Talk UI that is embedded in the public share authentication page; the template `publicshareauth` of Talk is appended to the one in the server when needed (that is, when sending the password by Talk is enabled for the share), and that template provides some JavaScript code that, when run on the browser, adjusts the page generated by the server to inject the Talk UI as needed.

request-password-by-talk

Currently the embedded Talk UI just provides chat; this will be improved in a different pull request with call support. Due to how the UI is injected this can be done without requiring any changes in the server, only in the Talk app. 

Also note that there are other shortcomings, like no information about whether the sharer has joined the room or not, or not being able to open and close the sidebar at will, which causes them to always overlap with the password input field during the conversation on narrow screens. For the time being it is just a basic UI... but at least it works :-)

Rooms can now be associated with arbitrary objects using an object type and an object id. This makes possible to provide custom behaviour for rooms based on that type using hooks to, for example, destroy the room as soon as one of the participants leave it or prevent extra users from joining the room.

The "share:password" rooms are a bit special, as they are one-time rooms created by one user, the receiver of the share, in behalf of another user, the sharer, who then becomes the owner of the room. In standard rooms no notification is sent to the owner when a room is created, so a special notification was added for this case. Depending on which one gets merged first, this pull request or #1029 will need to be updated.

@pygi @ivansss The notification works without problems in the WebUI, but I do not know if you may need to change something in the mobile apps to handle it. It would be good too to improve the message to include the name of the file or something like that, but that is something for another pull request too ;-)

### CODE REVIEW

**Comment by nickvergessen** [comment]
Why did you make it so complicated and coded talk into the Ui?
Just do a new window with a special public room. Free video support 🙈

File: js/publicshareauth.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+						.addClass('icon-loading-small-dark');
+
+				self.requestPassword();
+			});
+		},
+
+		setupRequestPasswordButton: function() {
+			// "submit-wrapper" is used to mimic the login button and thus get
+			// automatic colouring of the confirm icon by the Theming app
+			$('main').append('<div id="submit-wrapper" class="request-password-wrapper">' +
```
Comment (nickvergessen) [inline js/publicshareauth.js]: missing the `.` before `main` ?

Comment (danxuliu) [reply js/publicshareauth.js]: No, it refers to [an element](https://github.com/nextcloud/server/blob/7b848368d52ea84f3ce8e8b8792263de9f394c60/core/templates/layout.guest.php#L40-L42), not to a class.

**Comment by MorrisJobke** [comment]
@nickvergessen Feel free to merge - I will merge the server PR now.

File: lib/Migration/Version4000Date20180710162342.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+		$table = $schema->getTable('talk_rooms');
+
+		if (!$table->hasColumn('object_type')) {
+			$table->addColumn('object_type', Type::STRING, [
+				'notnull' => true,
+				'length' => 64,
+				'default' => '',
+			]);
+			$table->addColumn('object_id', Type::STRING, [
+				'notnull' => true,
```
Comment (nickvergessen) [inline lib/Migration/Version4000Date20180710162342.php]: false

File: lib/Migration/Version4000Date20180710162342.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+	public function changeSchema(IOutput $output, \Closure $schemaClosure, array $options) {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$table = $schema->getTable('talk_rooms');
+
+		if (!$table->hasColumn('object_type')) {
+			$table->addColumn('object_type', Type::STRING, [
+				'notnull' => true,
```
Comment (nickvergessen) [inline lib/Migration/Version4000Date20180710162342.php]: false

**Comment by nickvergessen** [comment]
@danxuliu I added some more commits, please have a look.
As requested by frank we are opening a new tab for now where you can do a video call to verify the identity before giving the password. This also has the advantage that the call survives your try to enter the password and you can tell the owner that the password worked.

I'm fine with merging it like that for now. If you manage to get calling in the "side"bar working we can simply revert the last commit and are back to the JS mode.

**Comment by nickvergessen** [comment]
Rebased to solve conlicts

Please review @danxuliu @Ivansss

File: lib/Migration/Version3003Date20180720162342.php [hunk 4]
```diff
@@ -39,12 +39,12 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
 
 		if (!$table->hasColumn('object_type')) {
 			$table->addColumn('object_type', Type::STRING, [
-				'notnull' => true,
+				'notnull' => false,
```
Comment (danxuliu) [inline lib/Migration/Version3003Date20180720162342.php]: Why this change?

**Comment by danxuliu** [comment]
I have cleaned up a bit the commit history, fixed some issues with the layout due to the changes in the server, and made the _Request password_ link look like a button.

File: lib/Migration/Version3003Date20180720162342.php (hunk 4 above)
Comment (danxuliu) [inline lib/Migration/Version3003Date20180720162342.php]: Why this change? Were not empty strings allowed already (in fact the default value is an empty string)?

File: lib/Migration/Version3003Date20180720162342.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+	public function changeSchema(IOutput $output, \Closure $schemaClosure, array $options) {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$table = $schema->getTable('talk_rooms');
+
+		if (!$table->hasColumn('object_type')) {
+			$table->addColumn('object_type', Type::STRING, [
+				'notnull' => false,
```
Comment (danxuliu) [inline lib/Migration/Version3003Date20180720162342.php]: Why this change? Were not empty strings allowed already (in fact the default value is an empty string)?

Comment (nickvergessen) [reply lib/Migration/Version3003Date20180720162342.php]: https://github.com/nextcloud/activity/commit/34279aed14493d167f2f50bfae1797048cbaac2c

All hail to oracle, storing empty string as null which then fails.

Comment (danxuliu) [reply lib/Migration/Version3003Date20180720162342.php]: Urg... Thanks for the explanation, I added it to the commit message for future reference.
